Isolatitude HEALPix padding

yikwill opened the pull request:
<!-- markdownlint-disable MD013-->
# Modulus Pull Request

## Description
<!-- Provide a standalone description of changes in this PR. -->
<!-- Reference any issues closed by this PR with "closes #1234". -->
<!-- Note: The pull request title will be included in the CHANGELOG. -->
This PR adds a new "isolatitude" padding strategy for applying 2D convolutions on the HEALPix grid. The name comes from the fact that this padding strategy will preserve zonally uniform (no east-west variance) signals everywhere except at the poles. The default strategy from Karlbauer et al. (2024) breaks zonal uniformity of incoming signals at the eight intersection points of three HEALPix faces (four such points at 45N and four at 45S) as well as along the seams where north (south) HEALPix faces meet another north (south) HEALPix face. The isolatitude padding scheme is illustrated below for an equatorial and northern hemisphere face.

![isolat_padding_E_face](https://github.com/user-attachments/assets/3a541543-5f28-4c00-9fbc-d60ff44fff1d)
![isolat_padding_N_face](https://github.com/user-attachments/assets/91ebd3f6-664c-475a-bbe8-39b1def95450)


## Checklist

- [x] I am familiar with the [Contributing Guidelines](https://github.com/NVIDIA/modulus/blob/main/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.
- [ ] The [CHANGELOG.md](https://github.com/NVIDIA/modulus/blob/main/CHANGELOG.md) is up to date with these changes.
- [ ] An [issue](https://github.com/NVIDIA/modulus/issues) is linked to this pull request.

## Dependencies

<!-- Call out any new dependencies needed if any -->

yikwill committed:
Initial commit

yikwill committed:
Adding isolatitude strategy for HPX padding

yikwill committed:
Adds padding mode option to model

yikwill committed:
Bug fixes

yikwill committed:
Removes debugging comments

akshaysubr commented on physicsnemo/models/dlwp_healpix/HEALPixRecUNet.py:
```diff
@@ -68,6 +68,7 @@ def __init__(
         enable_healpixpad: bool = False,
         couplings: list = [],
         residual_prediction: bool = True,
+        hpx_padding_mode: str = 'karlbauer',
```
Would be good to document this parameter and all possible options in the docstring below since this is the main entrypoint for a user

yikwill replied:
Good point, added.

akshaysubr commented on physicsnemo/models/dlwp_healpix_layers/healpix_layers.py:
```diff
@@ -604,7 +880,13 @@ def __init__(self, layer, **kwargs):
                 # or if this needs to be removed
                 layers.append(HEALPixPaddingv2(padding=padding))
```
Shouldn't there be an additional condition in the `if` branch here? Otherwise, if I set `enable_healpixpad=True` and `hpx_padding_mode=isolat`, I'll silently get the `karlbauer` padding right?

yikwill replied:
Good point. I've set it so that an error is thrown if `enable_healpixpad=True` and `hpx_padding_mode != "karlbauer"` since I think we want it to be clear that e2grid only supports Karlbauer padding for now. In the future when we have an optimized version of isolatitude padding, we can port it to e2grid and remove this check.

akshaysubr commented on physicsnemo/models/dlwp_healpix_layers/healpix_layers.py:
```diff
@@ -190,7 +189,12 @@ class HEALPixPadding(th.nn.Module):
+        self,
+        padding: int,
+        enable_nhwc: bool = False,
+        hpx_padding_mode: str = "karlbauer",
```
Not strictly needed for this PR, but it would be great if this module can be compatible with `torch.compile`. Either in this PR or later, would be good to make sure this compiles without errors and benchmark to see how much of the perf gap to the CUDA version can be made up.

yikwill replied:
I agree that it would be good to optimize this padding strategy. At the moment, I believe it is not compatible with `torch.compile`. I have yet to carefully profile the speed of isolatitude padding vs e2grid and non-e2grid Karlbauer padding. In the near future, I will do this and open a new PR to make this padding `torch.compile` compatible.

akshaysubr reviewed (changes requested):
Looks good to me. Only major comment is about some logic that can silently change the padding method.

yikwill committed:
Merge with dev

yikwill committed:
Adds additional documentation and checks

yikwill committed:
Adds hpx_padding_mode option to non-recurrent unet

yikwill committed:
Updates error message

yikwill commented:
Thanks for the review. Addressed the concerns and added the `hpx_padding_mode` option to the non-recurrent unet as well. I will save the `torch.compile` compatibility issue for another PR focused on performance optimization, but I agree that it is a near term priority.

nathanielcresswellclay reviewed (commented):
Sorry I'm so late to this. At this point we should probably do a rebase (or merge dev) to update this branch and make sure everything is still running smoothly. 

Can you also point me to a config in NV-dlesm so I can run some inference/training trials?

yikwill committed:
Merge branch 'dev' into wy/isolatitude_hpx_pad

yikwill commented:
No worries. Just merged dev. To avoid clogging the repo with more configs, here's a small config that you could run to test the padding scheme out.

```
defaults:
  - model: double_conv_256_128_128_couple
  - data: era5_hpx64_30var_coupled_sst_6h_24h_perlmutter
  - trainer: 30var_ws_uncertainty_weighted_adamw
  - _self_
hydra.job.chdir: true # to avoid have logs stepping on one another
experiment_name: isolat_test
output_dir: ./isolat_test
checkpoint_name: last
load_weights_only: false
seed: 0
batch_size: 1
learning_rate: 6e-4
num_workers: 8
port: 29450
master_address: localhost
hydra:
  verbose: false
  run:
    dir: ${output_dir}
model:
  hpx_padding_mode: isolat
  enable_nhwc: true
  enable_healpixpad: false
```

yikwill committed:
Merge branch 'dev' into wy/isolatitude_hpx_pad

yikwill committed:
Fixing mistakes from merge

yikwill committed:
Update error message

yikwill committed:
Adds capability to compile padding layers

yikwill committed:
Revert "Adds capability to compile padding layers"

This reverts commit 0c3e50099e6b0972758bb9b1f73a8bd0d6c0be86.

yikwill committed:
First attempt at major refactor to speed up isolatitude padding

yikwill committed:
Adds speed benchmarks

yikwill committed:
Adds earth2grid padding backend option

yikwill committed:
Merge branch 'dev' into wy/isolatitude_hpx_pad_refactor

yikwill committed:
Makes isolatitude padding torch compile compatible

yikwill committed:
Revert "Makes isolatitude padding torch compile compatible"

This reverts commit 7fb329bbbea3494cac5469ed3c3387efbe73ed8b.

yikwill committed:
Updates throughout the stack to make isolatitude padding compile compatible

yikwill committed:
Removes unnecessary options

yikwill committed:
Code cleanup

yikwill committed:
Removes isolatitude_reference padding option and cleans up code

yikwill committed:
Removing nvtx statements for torch compile compatibility

yikwill committed:
Removing compile wrapper

yikwill committed:
Forces internal fp32 computation

yikwill committed:
Removes benchmarking scripts

yikwill committed:
Documentation updates

yikwill committed:
Docstring updates

yikwill committed:
Update tests

yikwill commented:
The padding code has undergone a major refactor. Notably,
- The individual padding schemes have been moved to a new file called `healpix_paddings.py`. `healpix_layers.py` now only contains the module which wraps the given layer with hpx padding.
- The `enable_healpixpad` model option is now deprecated in favor of a new option called `hpx_padding_mode`. The available modes are `earth2grid`, `karlbauer`, and `isolatitude`. `hpx_padding_mode=earth2grid` produces the same behavior as `enable_healpixpad=True`. For backwards compatibility, I have left `enable_healpixpad` as an option in the UNets, but added a deprecation warning.
- A new model option `compile_padding` has been added which if true wraps the padding layer with torch compile.

yikwill committed:
Merge branch 'dev' into wy/isolatitude_hpx_pad

yikwill committed:
Updates patched energy score loss to be compatible with padding refactor

yikwill committed:
Ensures dlwp_healpix unet tests can run in isolation of other physicsnemo tests

yikwill committed:
Make tests skip invalid earth2grid padding but no cuda case

yikwill committed:
Merge branch 'wy/fix_dlwp_tests' into wy/isolatitude_hpx_pad

akshaysubr commented:
@yikwill With this deprecated `enable_healpixpad=True` option, can you also list how an existing training config that uses that option should be changed to still use the fast padding kernels?

yikwill committed:
Refactor isolatitude padding to account for sparse active positions in second gather

yikwill committed:
Adds tests for healpix paddings

yikwill committed:
Update deprecation warning with instructions on how to keep using earth2grid padding

yikwill committed:
Update deprecation warning and resolve correct hpx_padding_mode from enable_healpixpad

yikwill commented:
@akshaysubr Good idea. I've updated the deprecation warning with instructions on how to set `hpx_padding_mode` to replicate the behavior of the legacy option `enable_healpixpad=True`. I've added code to resolve the the correct `hpx_padding_mode`. If no `hpx_padding_mode` is provided and `enable_healpixpad=True`, the deprecation warning will be thrown and `hpx_padding_mode` will be set to `earth2grid`. If no `hpx_padding_mode` is provided and `enable_healpixpad=False`, the deprecation warning will be thrown and `hpx_padding_mode` will be set to `karlbauer`. Thus, the user will get the same behavior no matter what the value of `enable_healpixpad` is in their config. If `hpx_padding_mode` is explicitly provided, it will be used regardless of the value of `enable_healpixpad`.

Do you think there are any other places/ways to document this deprecation?

daviddpruitt commented on physicsnemo/metrics/climate/healpix_loss.py:
```diff
@@ -1012,8 +1011,9 @@ def __init__(
-        use_earth2grid_padding: bool = True,
-        enable_nhwc: bool = True,
+        hpx_padding_mode: str = "earth2grid",
+        enable_nhwc: bool = False,
```
What's the reason for switching this to false? Normally we get some speedup using nhwc.

yikwill replied:
We previously had a check that `enable_nhwc` had to be `False` if `enable_healpixpad=True`. However, there is no reason we couldn't have `earth2grid` padding support channels last. I've added `enable_nhwc` as an option to `HEALPixPaddingv2` and updated the default here to be `True`.

daviddpruitt commented on physicsnemo/metrics/climate/healpix_loss.py:
```diff
@@ -1032,10 +1032,12 @@ def __init__(
-        use_earth2grid_padding: bool
-            whether to use earth2grid hpx padding. Default is False.
+        hpx_padding_mode: str
+            HPX padding scheme to use. Default is "earth2grid".
```
Can you add the list of valid options here or in a config, that way we know what's available and what the options do.

yikwill replied:
I've updated the docstring to include the valid options and point the user to `healpix_paddings.py` if they want more info.

daviddpruitt commented on physicsnemo/models/dlwp_healpix_layers/healpix_layers.py:
```diff
@@ -15,592 +15,139 @@
+            elif hpx_padding_mode == "karlbauer":
+                padding_layer = HEALPixPadding(padding=padding, enable_nhwc=enable_nhwc)
+            elif hpx_padding_mode == "isolatitude":
+                padding_layer =  HEALPixPaddingIsolatitude(
```
This should have a check to make sure `nside` is passed in if `isolatitude` padding is selected

yikwill replied:
Added.

daviddpruitt commented on physicsnemo/metrics/climate/healpix_loss.py:
```diff
@@ -1082,12 +1085,28 @@ def __init__(
         if self.patch_radius > 0:
-            if self.use_earth2grid_padding:
+            if self.hpx_padding_mode == "earth2grid":
+                if not have_earth2grid or not th.cuda.is_available() or self.enable_nhwc:
```
Can this check be pushed into `healpix_paddings.py`? It would make more sense there since this is specific to paddings and it would allow other code to run the same check.

yikwill replied:
I added a function `make_hpx_padding_layer` to `healpix_paddings.py` that returns the correct padding layer given `hpx_padding_mode`, `enable_nhwc`, etc. Both `HEALPixLayer` and `PatchedEnergyScoreLoss` now use this helper function to get the right padding layer.

yikwill committed:
Adds channels last compatibility to earth2grid padding

yikwill committed:
Clarify padding scheme options in energy score loss docstring

yikwill committed:
Add check that nside is specified when using isolatitude padding

yikwill committed:
Adds make_healpix_layer helper function

yikwill committed:
Fix CUDA graph error

yikwill committed:
Merge branch 'dev' into wy/isolatitude_hpx_pad

yikwill committed:
Disable native torch conv padding

yikwill committed:
Make sure native conv padding is disable correctly

yikwill committed:
Merge branch 'dev' into wy/isolatitude_hpx_pad

yikwill committed:
Updated SmoothedInterpolateConv block to have same options as other blocks

yikwill committed:
Merge branch 'dev' into wy/isolatitude_hpx_pad

yikwill committed:
Merge branch 'dev' into wy/isolatitude_hpx_pad

yikwill committed:
Merge branch 'dev' into wy/isolatitude_hpx_pad

yikwill committed:
Merge branch 'dev' into wy/isolatitude_hpx_pad (dry_air_mass_constraint #37 landed).

yikwill committed:
Merge branch 'dev' into wy/isolatitude_hpx_pad (mainline sync).

yikwill committed:
Merge branch 'dev' into wy/isolatitude_hpx_pad (mainline sync).
